File: CommandPython/GarageDoorCommand.py
# Copyright (c) 2022 RodKingroo
from Command import *
from GarageDoor import *

# Garage Door Open Command
class GarageDoorOpenCommand(Command):
    def __init__ (self, garageDoor):
        if isinstance(garageDoor, GarageDoor):
            self.garageDoor = garageDoor
        else:
            raise NotImplementedError
        
    def execute(self):
        self.garageDoor.open()
        
    def undo(self):
        self.garageDoor.close()
  
# Garage Door Close Command      
class GarageDoorCloseCommand(Command):
    def __init__ (self, garageDoor):
        if isinstance(garageDoor, GarageDoor):
            self.garageDoor = garageDoor
        else:
            raise NotImplementedError
        
    def execute(self):
        self.garageDoor.close()
        
    def undo(self):
        self.garageDoor.open()

File: CommandCPP/GarageDoorCommand.cpp
/* Copyright (c) 2022 RodKingroo */
#include "GarageDoorCommand.h"

GarageDoorOpenCommand::GarageDoorOpenCommand(shared_ptr<GarageDoor> garageDoor)
{
    this->garageDoor = garageDoor;
}

void GarageDoorOpenCommand::execute() /* override */
{
    garageDoor->open();
}

void GarageDoorOpenCommand::undo() /* override */
{
    garageDoor->close();
}

GarageDoorCloseCommand::GarageDoorCloseCommand(shared_ptr<GarageDoor> garageDoor)
{
    this->garageDoor = garageDoor;
}

void GarageDoorCloseCommand::execute() /* override */
{
    garageDoor->close();
}

void GarageDoorCloseCommand::undo() /* override */
{
    garageDoor->open();
}
